packages/preview-middleware/src/types/index.ts

```diff
@@ -53,6 +53,11 @@ export interface FlpConfig {
     libs?: boolean;
     apps: App[];
     theme?: string;
+    test?: {
```

Comment:
We also need to add an optional `test/path` property since the middleware needs to expose both a simple FLP for previews as well as additional ones for testing. I would also assume that test should become an array because I need multiple different test html files if I want to test with QUnit and with Opa.

Reply:
Up to now I just created multiple entries in the yaml for testing. Might not be the best approach but it worked 🙈 

```
    - name: preview-middleware
      afterMiddleware: compression
      configuration:
        flp:
          path: localService/index.html
        debug: true
    - name: preview-middleware
      afterMiddleware: compression
      configuration:
        flp:
          path: test/integration/opaTests.qunit.html
          customInit: com/sap/cap/fe/ts/sample/test/integration/opaTests.qunit
          test:
            framework: "OPA5"
        debug: true
    - name: preview-middleware
      afterMiddleware: compression
      configuration:
        flp:
          path: test/unit/unitTests.qunit.html
          customInit: com/sap/cap/fe/ts/sample/test/unit/unitTests.qunit
          test:
            framework: "QUNIT"
        debug: true
```
How to deal with test/path (framework) dependent customInit in case everything is part of one configuration? Add it to the `test` as well?
```
    - name: preview-middleware
      afterMiddleware: compression
      configuration:
        flp:
          path: localService/index.html
          customInit: com/sap/cap/fe/ts/sample/localService/myCustomInit
          test:
            - framework: "OPA5"
              path: test/integration/opaTests.qunit.html
              customInit: com/sap/cap/fe/ts/sample/test/integration/opaTests.qunit
            - framework: "QUNIT"
              path: test/unit/unitTests.qunit.html
              customInit: com/sap/cap/fe/ts/sample/test/unit/unitTests.qunit
        debug: true
```

Reply:
Hm. I'm afraid this will get quite complex in case all those paths are part of one configuration.

Given the following configuration:

```
    - name: preview-middleware
      afterMiddleware: compression
      configuration:
        flp:
          path: localService/index.html
          customInit: com/sap/cap/fe/ts/sample/localService/myCustomInit
          test:
            - framework: "OPA5"
              path: test/integration/opaTests.qunit.html
              customInit: /test/integration/opaTests.qunit
            - framework: "QUNIT"
              path: test/unit/unitTests.qunit.html
              customInit: /test/unit/unitTests.qunit
        debug: true
```

will lead to
- flp.path -> adds a resource root (based on the app id) for `com.sap.cap.fe.ts.sample` path `..`
- flp.test.path -> defines its own prefixed resource root `open.ux.preview.test` (name is subject to change) path `../..` for the customInit to work, because `com.sap.cap.fe.ts.sample` is already in use and points to the "wrong" path (alternative: costomInit must always be located next to the html file from path)

So far so good. But now it gets worse

In case you use `paths` in `.tsconfig` it could look as follows:
- `"com/sap/cap/fe/ts/sample/*": ["./webapp/*"]`

Your `customInit` than can import based on the path:
`import("com/sap/cap/fe/ts/sample/test/unit/controller/CommentsSectionControllerTest")`

This will work at designtime and crash at runtime because the resource root for `com.sap.cap.fe.ts.sample` has been overwritten to `..`.

It would work if `open.ux.preview.test` is added to `.tsconfig` as path and used in the import statement of the `customInit`:
`import("open/ux/preview/test/test/unit/controller/CommentsSectionControllerTest")`

But this still can crash at runtime in case the tests themselves (`CommentsSectionControllerTest`) or any of their dependencies (`CommentsSectionController`) use imports based on the overwritten resource root path `com.sap.cap.fe.ts.sample`, e.g. `import { Message, Model, Ui } from "com/sap/cap/fe/ts/sample/ext/utils/Constants";`

Being not able to use `paths` in `.tsconfig` for preview-middleware to work seems quite annoying. @tobiasqueck do you have any idea how to work around this? (I did not yet commit these changes but I can of course do so if it helps)

The issue does not occur with multiple preview-middleware instances for each path because this ensures the resource root for the app id is always correct.

Reply:
Thansk a ton for the feedback. I mainly tested the idea with js files but will add a ts project to my testing.
Let me tinker a little with it and then come back to you.

Reply:
Yeah, sorry for the lengthy comment 🙈 
I just committed something that at least forks for my test setup. Might be a good idea to continue from there with further adjustments. Needed to slightly adjust the config (test.customInit needs to use the app id to match the resource root):

```
    - name: preview-middleware
      afterMiddleware: compression
      configuration:
        flp:
          path: localService/index.html
          test:
            - framework: "OPA5"
              path: test/integration/opaTests.qunit.html
              customInit: com/sap/cap/fe/ts/sample/test/integration/opaTests.qunit
            - framework: "QUNIT"
              path: test/unit/unitTests.qunit.html
              customInit: com/sap/cap/fe/ts/sample/test/unit/unitTests.qunit
        debug: true
```

Reply:
yes, agree

Reply:
actually... doing it the other way around seems to be much easier as the express router.get needs to be instantiated without ns whereas the module.require needs a ns. So as we anyway have have the ns at hand the user just needs to provide the relative path from webapp and we can do the rest. That fits well to the relative `path` property too. What do you think?

I just tested with
```
    - name: preview-middleware
      afterMiddleware: compression
      configuration:
        flp:
          path: localService/index.html
        test:
          - framework: "OPA5"
            path: /test/integration/opaTests.qunit.html
            init: /test/integration/opaTests.qunit
          - framework: "Qunit"
        debug: true
```
and it worked pretty well.